Clearify doSim and doSimUntilVoid use

andreasWallner opened the pull request.

Dolu1990 commented:
Hmmm 
"When in doubt about what to use, prefer ``doSim(...)``, for example"

personnaly i'm kind of always using doSimUntilVoid.
I mean, maybe instead we could provide a very minimal typical testbench for each ?

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -79,18 +79,18 @@ This function will return a ``SimCompiled`` instance.
+``doSim[(simName[, seed])]{dut => /* main stimulus code */}``
+  Run the simulation until the main thread runs to completion and exits/returns.
+  It will detect and report an error if the simulation gets fully stuck. As long as
+  a e.g. a clock is running the simulation can continue forever, it is therefore recommended
```
```suggestion
  e.g. a clock is running the simulation can continue forever, it is therefore recommended
```

Repetition of a

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -79,18 +79,18 @@ This function will return a ``SimCompiled`` instance.
+  Run the simulation until the main thread runs to completion and exits/returns.
+  It will detect and report an error if the simulation gets fully stuck. As long as
+  a e.g. a clock is running the simulation can continue forever, it is therefore recommended
+  to use ``simTimeout(cycles)`` to limit the possible runtime.
```
```suggestion
  to use ``SimTimeout(cycles)`` to limit the possible runtime.
```

I think it is `SimTimeout` not `simTimeout`.

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -79,18 +79,18 @@ This function will return a ``SimCompiled`` instance.
 
+``doSimUntilVoid[(simName[, seed])]{dut => ...}``
+  Run the simulation until it is ended by calling either ``simSuccess()`` or ``simFailure()``.
+  The main stimulus thread can continue or exit early, as long as there are events to process,
```
I am not sure I understand: does `simSuccess` stop simulation or not?

andreasWallner replied:
`simSuccess` will stop the simulation

numero-744 replied:
So I find this sentence being just after the end functions confusing "The main stimulus thread can continue or exit early".

What about telling that another way?

* the end of the block does not stop simulation
* simulation stops only when no more events to process, or called one of the two end functions

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -103,6 +103,7 @@ For example :
      .workspacePath("~/tmp")
      .compile(new TopLevel)
      .doSim { dut =>
+       simTimeout(1000)
```
```suggestion
       SimTimeout(1000)
```

andreasWallner commented:
> personnaly i'm kind of always using doSimUntilVoid.

Interesting, I had never used it before quickly trying it when I wrote about it... (also: in the SpinalHDL repo there are 152 usages of doSim and 48 of doSimUntilVoid ;-))
I'll add a sample for `doSimUntilVoid` as well

Dolu1990 commented:
@andreasWallner 

May this example help : ? 

```scala
object SimMiaouDoc extends App {
  class TopLevel extends Component {
    val counter = out(Reg(UInt(8 bits)) init (0))
    counter := counter + 1
  }

  SimConfig.compile(new TopLevel).doSim { dut =>
    SimTimeout(1000)
    dut.clockDomain.forkStimulus(10)
    dut.clockDomain.waitSamplingWhere(dut.counter.toInt == 20)
  }

  SimConfig.compile(new TopLevel).doSimUntilVoid{ dut =>
    SimTimeout(1000)
    dut.clockDomain.forkStimulus(10)
    fork {
      dut.clockDomain.waitSamplingWhere(dut.counter.toInt == 20)
      simSuccess()
    }
  }
}
```

numero-744 commented:
With these examples, maybe a note explaining that with `doSim` the second example would not wait*?

*: Err… wait is not really visible, maybe adding a `println("done")` before `simSuccess()` and tell it does not print if using `doSim` instead of `doSimUntilVoid`?

Dolu1990 commented:
@numero-744 

> *: Err… wait is not really visible, maybe adding a println("done") before simSuccess() and tell it does not print if using doSim instead of doSimUntilVoid?

Ahhh sure ^^, why not :)

Dolu1990 commented:
@andreasWallner Would that be good for you ?

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -79,18 +79,18 @@ This function will return a ``SimCompiled`` instance.
+``doSimUntilVoid[(simName[, seed])]{dut => ...}``
+  Run the simulation until it is ended by calling either ``simSuccess()`` or ``simFailure()``.
+  The main stimulus thread can continue or exit early, as long as there are events to process,
+  the simulation will continue. The simulation will report an error if it gets fully stuck.
```
I think there is an issue with the code: wouldn't a function named `doSimUntilVoid` be expected to end successfully when void is reach / there are no events to process anymore?

Dolu1990 replied:
That was the initial intend, but it is kinda never used at all and not usefull, and instead provide some easy to produce false SimSuccess.
So we turned it into SimFailure if the simulation has nothing scheduled anymore.

numero-744 replied:
So the name did match the behavior, but it does not anymore… So the name does not match the behavior!

Dolu1990 commented:
I added the example :)

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -87,13 +87,22 @@ On this ``SimCompiled`` instance you can run your simulation with the following
+      counter := counter + 1
+   }
+
+Here is a template with many simulation configuration :
```
Typos

```
Here is a template with many simulation configurations:
```

Dolu1990 replied:
Hmm i would say fondamentaly, doSimUntilVoid name do not specify anything about what hapen once void is here.

Typo fixed ^^

numero-744 replied:
> fondamentaly, `doSimUntilVoid` name do not specify anything about what hapen once void is here.

You got the point :smile:

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -107,6 +116,45 @@ When in doubt about what to use, prefer ``doSim(...)``, for example :
        // Simulation code here
    }
 
+Here is a template where the simulation end by completting the simulation main thread execution :
```
```
Here is a template where the simulation ends by completing the simulation main thread execution:
```

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -107,6 +116,45 @@ When in doubt about what to use, prefer ``doSim(...)``, for example :
+      println("done")
+    }
+    
+Here is a template where the simulation end by explicitly calling a simSuccess() :
```
```
Here is a template where the simulation ends by explicitly calling `simSuccess()`:
```

numero-744 commented on source/SpinalHDL/Simulation/bootstraps.rst:
```diff
@@ -107,6 +116,45 @@ When in doubt about what to use, prefer ``doSim(...)``, for example :
+      }
+    }
+
+Note is it equivalent to : 
```
```
Note is it equivalent to:

```

numero-744 replied:
Missing empty line here provokes an error, which was reported but was ignored and the branch was merged. I think this merge was premature, and waiting for a review (even if you are the manager) would have avoided that.

See just above this title: https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/bootstraps.html#running-multiple-tests-on-the-same-hardware

numero-744 commented:
Oh I forgot to click "update" on one of my comments above, sorry ^^'

I had suggested this: https://github.com/andreasWallner/SpinalDoc-RTD/pull/2/files

Btw, I liked the example you used initially because it was extending `App`, which makes it easy to test quickly by copy-pasting it, but it it true that it does not let us add comments as text between lines.

numero-744 commented:
I understand

andreasWallner commented:
Sorry for the silence, I was quite sick the last few days, I'll have a look later

Dolu1990 commented:
No worries ^^

andreasWallner commented:
I had a quick look and just changed one of the sentences and fixed a missing newline that yielded a broken output. Looks OK from my point of view but: I'm not sure the examples make one appreciate the differences or advantages of the approaches - mostly because there is only a single line difference.
I was originally thinking to link to the examples that are already there in the documentation and are a good showcase, namely https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/examples/asynchronous.html and https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/examples/single_clock_fifo.html and if needed add a few words there why the approach was chosen. What do you guys think about that?

andreasWallner commented:
Sorry - It took me until now until I saw that I was discussing on a merged PR...

Dolu1990 commented:
@andreasWallner  :D
So merged stuff doesn't mean can't be PR again for new changes :D
Nothing is definitive.

So, i think you are right, having link to the example may be a good thing instead of stuffing that section with too many.
What's about moving the added example into a new example section with more details ? Or yes, as you said, https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/examples/asynchronous.html and https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/examples/single_clock_fifo.html with some alternatives.

A PR can be open :)
